File: ainpc-core-plugin/src/main/java/ro/ainpc/spawn/NpcSpawnPlan.java
package ro.ainpc.spawn;

public record NpcSpawnPlan(
    String npcKey,
    String name,
    String occupation,
    String backstory,
    int age,
    String gender,
    String archetype,
    String homePlaceId,
    String workPlaceId,
    String socialPlaceId,
    String spawnNodeId,
    String homeNodeId,
    String workNodeId,
    String socialNodeId,
    String familyId
) {
    public NpcSpawnPlan {
        npcKey = clean(npcKey);
        name = clean(name);
        occupation = clean(occupation);
        backstory = clean(backstory);
        age = age > 0 ? age : 30;
        gender = clean(gender).isBlank() ? "male" : clean(gender).toLowerCase();
        archetype = clean(archetype);
        homePlaceId = clean(homePlaceId);
        workPlaceId = clean(workPlaceId);
        socialPlaceId = clean(socialPlaceId);
        spawnNodeId = clean(spawnNodeId);
        homeNodeId = clean(homeNodeId);
        workNodeId = clean(workNodeId);
        socialNodeId = clean(socialNodeId);
        familyId = clean(familyId);
    }

    public static Builder builder(String npcKey, String name) {
        return new Builder(npcKey, name);
    }

    public String sourceKey() {
        String scope = firstNonBlank(familyId, homePlaceId, spawnNodeId, "global");
        String key = firstNonBlank(npcKey, name, "npc");
        return "spawn_plan:" + normalizeToken(scope) + ":" + normalizeToken(key);
    }

    public static final class Builder {
        private final String npcKey;
        private final String name;
        private String occupation = "";
        private String backstory = "";
        private int age = 30;
        private String gender = "male";
        private String archetype = "";
        private String homePlaceId = "";
        private String workPlaceId = "";
        private String socialPlaceId = "";
        private String spawnNodeId = "";
        private String homeNodeId = "";
        private String workNodeId = "";
        private String socialNodeId = "";
        private String familyId = "";

        private Builder(String npcKey, String name) {
            this.npcKey = npcKey;
            this.name = name;
        }

        public Builder occupation(String occupation) {
            this.occupation = occupation;
            return this;
        }

        public Builder backstory(String backstory) {
            this.backstory = backstory;
            return this;
        }

        public Builder age(int age) {
            this.age = age;
            return this;
        }

        public Builder gender(String gender) {
            this.gender = gender;
            return this;
        }

        public Builder archetype(String archetype) {
            this.archetype = archetype;
            return this;
        }

        public Builder homePlaceId(String homePlaceId) {
            this.homePlaceId = homePlaceId;
            return this;
        }

        public Builder workPlaceId(String workPlaceId) {
            this.workPlaceId = workPlaceId;
            return this;
        }

        public Builder socialPlaceId(String socialPlaceId) {
            this.socialPlaceId = socialPlaceId;
            return this;
        }

        public Builder spawnNodeId(String spawnNodeId) {
            this.spawnNodeId = spawnNodeId;
            return this;
        }

        public Builder homeNodeId(String homeNodeId) {
            this.homeNodeId = homeNodeId;
            return this;
        }

        public Builder workNodeId(String workNodeId) {
            this.workNodeId = workNodeId;
            return this;
        }

        public Builder socialNodeId(String socialNodeId) {
            this.socialNodeId = socialNodeId;
            return this;
        }

        public Builder familyId(String familyId) {
            this.familyId = familyId;
            return this;
        }

        public NpcSpawnPlan build() {
            return new NpcSpawnPlan(
                npcKey,
                name,
                occupation,
                backstory,
                age,
                gender,
                archetype,
                homePlaceId,
                workPlaceId,
                socialPlaceId,
                spawnNodeId,
                homeNodeId,
                workNodeId,
                socialNodeId,
                familyId
            );
        }
    }

    private static String clean(String value) {
        return value == null ? "" : value.trim();
    }

    private static String firstNonBlank(String... values) {
        if (values == null) {
            return "";
        }
        for (String value : values) {
            String cleanValue = clean(value);
            if (!cleanValue.isBlank()) {
                return cleanValue;
            }
        }
        return "";
    }

    private static String normalizeToken(String value) {
        String cleaned = clean(value).toLowerCase(java.util.Locale.ROOT);
        if (cleaned.isBlank()) {
            return "unknown";
        }
        return cleaned.replaceAll("[^a-z0-9_.:-]+", "_");
    }
}

File: ainpc-core-plugin/src/main/kotlin/ro/ainpc/spawn/NpcSpawnPlan.kt
package ro.ainpc.spawn

import java.util.Locale

data class NpcSpawnPlan(
    val npcKey: String,
    val name: String,
    val occupation: String,
    val backstory: String,
    val age: Int,
    val gender: String,
    val archetype: String,
    val homePlaceId: String,
    val workPlaceId: String,
    val socialPlaceId: String,
    val spawnNodeId: String,
    val homeNodeId: String,
    val workNodeId: String,
    val socialNodeId: String,
    val familyId: String
) {
    private val cleanedNpcKey = clean(npcKey)
    private val cleanedName = clean(name)
    private val cleanedOccupation = clean(occupation)
    private val cleanedBackstory = clean(backstory)
    private val cleanedAge = if (age > 0) age else 30
    private val cleanedGender = clean(gender).let { if (it.isBlank()) "male" else it.lowercase(Locale.getDefault()) }
    private val cleanedArchetype = clean(archetype)
    private val cleanedHomePlaceId = clean(homePlaceId)
    private val cleanedWorkPlaceId = clean(workPlaceId)
    private val cleanedSocialPlaceId = clean(socialPlaceId)
    private val cleanedSpawnNodeId = clean(spawnNodeId)
    private val cleanedHomeNodeId = clean(homeNodeId)
    private val cleanedWorkNodeId = clean(workNodeId)
    private val cleanedSocialNodeId = clean(socialNodeId)
    private val cleanedFamilyId = clean(familyId)

    fun npcKey(): String = cleanedNpcKey
    fun name(): String = cleanedName
    fun occupation(): String = cleanedOccupation
    fun backstory(): String = cleanedBackstory
    fun age(): Int = cleanedAge
    fun gender(): String = cleanedGender
    fun archetype(): String = cleanedArchetype
    fun homePlaceId(): String = cleanedHomePlaceId
    fun workPlaceId(): String = cleanedWorkPlaceId
    fun socialPlaceId(): String = cleanedSocialPlaceId
    fun spawnNodeId(): String = cleanedSpawnNodeId
    fun homeNodeId(): String = cleanedHomeNodeId
    fun workNodeId(): String = cleanedWorkNodeId
    fun socialNodeId(): String = cleanedSocialNodeId
    fun familyId(): String = cleanedFamilyId

    fun sourceKey(): String {
        val scope = firstNonBlank(cleanedFamilyId, cleanedHomePlaceId, cleanedSpawnNodeId, "global")
        val key = firstNonBlank(cleanedNpcKey, cleanedName, "npc")
        return "spawn_plan:${normalizeToken(scope)}:${normalizeToken(key)}"
    }

    class Builder private constructor(
        private val npcKey: String,
        private val name: String
    ) {
        private var occupation = ""
        private var backstory = ""
        private var age = 30
        private var gender = "male"
        private var archetype = ""
        private var homePlaceId = ""
        private var workPlaceId = ""
        private var socialPlaceId = ""
        private var spawnNodeId = ""
        private var homeNodeId = ""
        private var workNodeId = ""
        private var socialNodeId = ""
        private var familyId = ""

        fun occupation(occupation: String): Builder {
            this.occupation = occupation
            return this
        }

        fun backstory(backstory: String): Builder {
            this.backstory = backstory
            return this
        }

        fun age(age: Int): Builder {
            this.age = age
            return this
        }

        fun gender(gender: String): Builder {
            this.gender = gender
            return this
        }

        fun archetype(archetype: String): Builder {
            this.archetype = archetype
            return this
        }

        fun homePlaceId(homePlaceId: String): Builder {
            this.homePlaceId = homePlaceId
            return this
        }

        fun workPlaceId(workPlaceId: String): Builder {
            this.workPlaceId = workPlaceId
            return this
        }

        fun socialPlaceId(socialPlaceId: String): Builder {
            this.socialPlaceId = socialPlaceId
            return this
        }

        fun spawnNodeId(spawnNodeId: String): Builder {
            this.spawnNodeId = spawnNodeId
            return this
        }

        fun homeNodeId(homeNodeId: String): Builder {
            this.homeNodeId = homeNodeId
            return this
        }

        fun workNodeId(workNodeId: String): Builder {
            this.workNodeId = workNodeId
            return this
        }

        fun socialNodeId(socialNodeId: String): Builder {
            this.socialNodeId = socialNodeId
            return this
        }

        fun familyId(familyId: String): Builder {
            this.familyId = familyId
            return this
        }

        fun build(): NpcSpawnPlan {
            return NpcSpawnPlan(
                npcKey,
                name,
                occupation,
                backstory,
                age,
                gender,
                archetype,
                homePlaceId,
                workPlaceId,
                socialPlaceId,
                spawnNodeId,
                homeNodeId,
                workNodeId,
                socialNodeId,
                familyId
            )
        }

        companion object {
            @JvmStatic
            fun create(npcKey: String, name: String): Builder = Builder(npcKey, name)
        }
    }

    companion object {
        @JvmStatic
        fun builder(npcKey: String, name: String): Builder = Builder.create(npcKey, name)

        private fun clean(value: String?): String = value?.trim() ?: ""

        private fun firstNonBlank(vararg values: String?): String {
            for (value in values) {
                val cleanValue = clean(value)
                if (cleanValue.isNotBlank()) {
                    return cleanValue
                }
            }
            return ""
        }

        private fun normalizeToken(value: String?): String {
            val cleaned = clean(value).lowercase(Locale.ROOT)
            if (cleaned.isBlank()) {
                return "unknown"
            }
            return cleaned.replace(Regex("[^a-z0-9_.:-]+"), "_")
        }
    }
}
